### PR TITLE
Feature/cambium ap modules

### PR DESCRIPTION
**Pull request by atran-inverse**
# Description
New Cambium AP module for 802.1x, MAC Auth and Web auth.
Added documentation

# Impacts
Allows PF to use Cambium AP

# Delete branch after merge
YES

### CODE REVIEW

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 1]
```diff
@@ -4701,6 +4701,64 @@ RF Switches
...
+^^^^^^^^^^^^
+
+802.1x
+++++++
+
+To setup the Cambium cnPilot E410 AP to use 802.1x, first, you need to already have configured the VLANs that will be used in the AP under *Configure -> Network*. Make sure that in *Configure -> Network -> Ethernet Ports*, the port is configured to *Trunk Multiple VLANs*, and the list of VLANs are allowed.
+
+Next, go to *Configure -> WLAN*, and click on *Add WLAN*. Give it the desired ID, and enter your SSID, default VLAN, and select *WPA2 Enterprise* for Security.
+
+image::docs/images/cambium-dot1x-basic.png[scaledwidth="100%",alt="Cambium WLAN Basic"]
```
Comment (cgx) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc:4717]: The image is missing.

File: lib/pf/Switch/Cambium.pm [hunk 2]
```diff
@@ -0,0 +1,207 @@
+package pf::Switch::Cambium;
+
+=head1 NAME
+
+pf::Switch::Cambium
+
+=head1 SYNOPSIS
+
+Implements a Cambium AP which supports 802.1x in wireless
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: mac-auth and webauth

File: lib/pf/Switch/Cambium.pm [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+# Description
+sub description { return "Cambium" }
+
+# CAPABILITIES
+# access technology supported
+sub supportsWirelessDot1x { return $TRUE; }
+sub supportsWirelessMacAuth { return $TRUE; }
+sub supportsExternalPortal { return $TRUE; }
+sub supportsWebFormRegistration { return $TRUE }
+
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: sub inlineCapabilities { return ($MAC,$PORT); }

File: lib/pf/Switch/Cambium.pm [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+    
+    %params = (
+        switch_id               => $req->param('ga_srvr'),
+        client_mac              => clean_mac($req->param('ga_cmac')),
+        client_ip               => defined($req->param('ga_cip')) ? $req->param('ga_cip') : undef,
+        ssid                    => $req->param('ga_ssid'),
+        synchronize_locationlog => $TRUE,
+    );
+    use Data::Dumper;
+    $logger->info(Dumper(\%params));
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: remove

File: lib/pf/Switch/Cambium.pm [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+
+Parse external portal request using URI and its parameters then return a hash reference with the appropriate parameters
+
+See L<pf::web::externalportal::handle>
+
+=cut
+
+sub parseExternalPortalRequest {
+    my ( $self, $r, $req ) = @_;
+    my $logger = $self->logger;
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: remove

File: lib/pf/Switch/Cambium.pm [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    my %params = ();
+    
+    %params = (
+        switch_id               => $req->param('ga_srvr'),
+        client_mac              => clean_mac($req->param('ga_cmac')),
+        client_ip               => defined($req->param('ga_cip')) ? $req->param('ga_cip') : undef,
+        ssid                    => $req->param('ga_ssid'),
+        synchronize_locationlog => $TRUE,
+    );
+    use Data::Dumper;
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: remove

File: lib/pf/Switch/Cambium.pm [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+use strict;
+use warnings;
+
+use POSIX;
+use Try::Tiny;
+
+use base ('pf::Switch');
+
+use pf::config qw(
+    $MAC
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

File: lib/pf/Switch/Cambium.pm [hunk 8]
```diff
@@ -0,0 +1,207 @@
...
+    $SSID
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

File: lib/pf/Switch/Cambium.pm [hunk 9]
```diff
@@ -0,0 +1,207 @@
...
+    $WIRELESS_MAC_AUTH
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

File: lib/pf/Switch/Cambium.pm [hunk 10]
```diff
@@ -0,0 +1,207 @@
...
+
+Nothing documented at this point.
+
+=cut
+
+use strict;
+use warnings;
+
+use POSIX;
+use Try::Tiny;
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

File: lib/pf/Switch/Cambium.pm [hunk 11]
```diff
@@ -0,0 +1,207 @@
...
+=head1 BUGS AND LIMITATIONS
+
+Nothing documented at this point.
+
+=cut
+
+use strict;
+use warnings;
+
+use POSIX;
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

File: lib/pf/Switch/Cambium.pm [hunk 12]
```diff
@@ -0,0 +1,207 @@
...
+
+use base ('pf::Switch');
+
+use pf::config qw(
+    $MAC
+    $SSID
+    $WIRELESS_MAC_AUTH
+);
+use pf::constants;
+use pf::node;
```
Comment (fdurand) [inline lib/pf/Switch/Cambium.pm]: useless

**Review by fdurand — CHANGES_REQUESTED** [review]

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 13]
```diff
@@ -4701,6 +4701,64 @@ RF Switches
 
 See the <<_motorola,Motorola RF Switches>> documentation.
 
+Cambium
+~~~~~~~
+
+cnPilot E410
+^^^^^^^^^^^^
+
+802.1x
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: 802.1x -> 802.1X everywhere

File: docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc [hunk 14]
```diff
@@ -4701,6 +4701,64 @@ RF Switches
...
+++++++
+
+To setup the Cambium cnPilot E410 AP to use 802.1x, first, you need to already have configured the VLANs that will be used in the AP under *Configure -> Network*. Make sure that in *Configure -> Network -> Ethernet Ports*, the port is configured to *Trunk Multiple VLANs*, and the list of VLANs are allowed.
+
+Next, go to *Configure -> WLAN*, and click on *Add WLAN*. Give it the desired ID, and enter your SSID, default VLAN, and select *WPA2 Enterprise* for Security.
+
+image::docs/images/cambium-dot1x-basic.png[scaledwidth="100%",alt="Cambium WLAN Basic"]
+
+In the *Radius Server* tab, enter the management IP address of PacketFence (VIP in case of a cluster) and the Radius secret for Authentication and Accounting servers.
```
Comment (extrafu) [inline docs/PacketFence_Network_Devices_Configuration_Guide.asciidoc]: Radius -> RADIUS

File: lib/pf/Switch/Cambium.pm [hunk 15]
```diff
@@ -0,0 +1,207 @@
...
+            <input type="hidden" name="ga_srvr" value="$srvr">
+            <input type="hidden" name="ga_cmac" value="$client_mac">
+            <input type="hidden" name="ga_cip" value="$client_ip">
+            <input type="hidden" name="ga_Qv" value="$qv">
+            <input type="hidden" name="ga_orig_url" value="$orig_url">
+        </form>
+        <script src="/content/autosubmit.js" type="text/javascript"></script>
+    ];
+
+    $logger->debug("Generated the following html form : ".$html_form);
```
Comment (extrafu) [inline lib/pf/Switch/Cambium.pm:159]: Remove?

Comment (fdurand) [reply lib/pf/Switch/Cambium.pm:159]: IMO no, it can be usefull

**Pull request merged**